src/bootman/bootman.c

```diff
@@ -446,6 +458,174 @@ bool boot_manager_set_uname(BootManager *self, const char *uname)
+                }
+
+                /* Some kind of broken link */
+                if (lstat(path, &st) != 0) {
+                        continue;
+                }
+
+                /* Regular only */
+                if (!S_ISREG(st.st_mode)) {
+                        continue;
+                }
+
+                /* empty files are skipped too */
+                if (st.st_size == 0) {
+                        continue;
+                }
+
+                initrd_name = strdup(ent->d_name);
+                if (!nc_array_add(initrd_nodep, initrd_name)) {
+                        DECLARE_OOM();
+                        abort();
+                }
+        }
+        if (initrd_nodep->len > 0) {
+                nc_array_qsort(initrd_nodep, initrd_nodep_name_cmp);
+                self->initrd_nodep = strdup(initrd_nodep->data[0]);
+        }
+        nc_array_free(&initrd_nodep, free);
+        closedir(dir);
+}
+
+void boot_manager_copy_initrd_nodep(BootManager *self)
+{
+        autofree(char) *base_path = NULL;
+        autofree(char) *initrd_target = NULL;
+        autofree(char) *initrd_source = NULL;
+        bool is_uefi = ((self->bootloader->get_capabilities(self) & BOOTLOADER_CAP_UEFI) ==
+                        BOOTLOADER_CAP_UEFI);
+        const char *efi_boot_dir =
+            is_uefi ? self->bootloader->get_kernel_destination(self) : NULL;
+        base_path = boot_manager_get_boot_dir((BootManager *)self);
+        if (!self || !self->initrd_nodep_dir || self->initrd_nodep == NULL) {
+                return;
+        }
+
+        /* if it's UEFI, then bootloader->get_kernel_dst() must return a value. */
+        if (is_uefi && !efi_boot_dir) {
+                return;
+        }
+
+        initrd_target = string_printf("%s%s/%s",
+                                      base_path,
+                                      (is_uefi ? efi_boot_dir : ""),
+                                      self->initrd_nodep);
+        initrd_source = string_printf("%s/%s",
+                                      self->initrd_nodep_dir,
+                                      self->initrd_nodep);
+        if (!cbm_files_match(initrd_source, initrd_target)) {
+                if (!copy_file_atomic(initrd_source, initrd_target, 00644)) {
+                        LOG_FATAL("Failed to install initrd %s: %s",
+                                  initrd_target,
+                                  strerror(errno));
+                        return;
+                }
+        }
+}
+
+void boot_manager_remove_initrd_nodep(BootManager * self)
+{
+        autofree(char) *base_path = NULL;
+        autofree(char) *initrd_dir = NULL;
+        autofree(char) *initrd_target = NULL;
+        bool is_uefi = ((self->bootloader->get_capabilities(self) & BOOTLOADER_CAP_UEFI) ==
+                        BOOTLOADER_CAP_UEFI);
+        const char *efi_boot_dir =
+            is_uefi ? self->bootloader->get_kernel_destination(self) : NULL;
+        DIR *dir = NULL;
```

Comment:
Can likely be an `autofree` candidate?

Reply:
I can add autofree for DIR *, should I do it just for my changes or in all the parts that use DIR * in the repo?